### ISSUE TITLE
On property queries and algorithms …

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#22140]
We might have missed the boat on this one, but it would be possible to include a property query are part of an algorithm name.
The concept is to introduce a reserved separator that splits the algorithm name string into the actual name and a property query to use to fetch it.

E.g. `AES-256-CGM#fips="yes"` would fetch the `AES-256-GCM` algorithm with a property query of `fips="yes"`.

The advantage is that it lets the query be carried along with the algorithm name.  This would allow different queries to be used for different algorithms within a single operation.

We can [bikeshed](https://en.wikipedia.org/wiki/Law_of_triviality) as to the separator character or string and the interaction between the property queries in the algorithm name, passed to the fetch call and the library context's global default query.

**Comment by levitte** [comment]
We talked about exactly this as part of the OTC f2f in June, so we're several who would be in favor of this sort of thing.

An idea I have in my head is that this sort of property would be a property *of the algorithm* rather than a property *of the implementation*...  it's a distinction we haven't thought of before ('cause we only have one type of property, obviously), but that might have some value.  

A reason I'm thinking of this is the usual topic that's been the base of these discussions, the granularity of EC keys.  Using your syntax, I'd say `EC#group=secp384r1` would be viable, for example.  And oh BTW, possible implementation bikeshed, but doesn't that indicate that this algorithm properties could be turned into an OSSL_PARAM element to be passed to the implementation, er, at some point?

While thinking about this, do we want to allow specifications of multiple algorithm properties?  What syntax should we have to allow that?

**Comment by paulidale** [comment]
I'm not sure OSSL_PARAM is a good way to make the distinction.  Algorithm name would be far better IMO.
Yes, we do need to resolve the EC curve issue & probably sooner than we anticipate.

Properties are comma separated for conjunction.  We have no capability for disjunction currently.  IMO, adding such requires proper expression parsing with parenthesis.  Not a big deal to implement but a big conceptual change and a far more complicated grammar.

I wasn't at the F2F and, although I attended the majority of session, I don't remember discussions about this.

**Comment by levitte** [comment]
> I'm not sure OSSL_PARAM is a good way to make the distinction. Algorithm name would be far better IMO.

... and each leading to its own `OSSL_DISPATCH`.  Yeah of course, that's workable, and goes along the lines of already existing ideas on declaring composite algorithms...  that should take care of all the setups that need to be done programmatically on the provider end.

**Comment by mattcaswell** [comment]
> This would allow different queries to be used for different algorithms within a single operation.

I'm not sure I understand what this means. Can you give an example?

**Comment by ifranzki** [comment]
Nice idea!  This could actually be an alternative to what I proposed in https://github.com/openssl/openssl/pull/21961#issuecomment-1721118765. 

With this I could for example  control that different algorithms used by the TLS protocol (i.e. for a SSL_CTX) use different providers, just by supplying a property query that selects different providers for different algos, e.g. `ECDSA#provider=aaaa,ECDH#provider=bbb` selecting provider aaa for ECDSA and provider bbb for ECDH.

**Comment by paulidale** [comment]
@ifranzki, exactly.

@mattcaswell, I'll try to come up with something.  A not very good example might be: `HMAC#provider?="default"` passing the algorithm `SHA2-256#fips="yes"`.  We can do the same thing currently via params, so it's not terribly meaningful.

**Comment by ifranzki** [comment]
A possible problem may arise with asymmetric algorithms. There to each algorithm/method, the corresponding keymgmt implementation of the same provider should be used. 

It sometimes does make a difference with which provider a key is generated. Take an HSM/PKCS#11 based provider: A (private) key generated with HSM/PKCS#11 will in most cases not be extractable, thus that (private) key can not not be transported (via export/import) from one provider to the other. So a key that will later be used with an algorithm provided by a clear key provider can for example not be generated with a HSM/PKCS#11 based provider, because you can't extract the clear key from the HSM/PKCS#11 based key, but that would be needed to export it and import it to the clear key provider. 

So we should somehow bind the key generation operation to the algorithms. 

Like with my example `ECDSA#provider=aaa,ECDH#provider=bbb`: If provider `aaa` is a PKCS#11 provider, and `bbb` is the default provider, then the key used for the ECDSA operation should be a PKCS#11 based key (possibly imported via STORE through an PKCS#11 uri), but the ephemeral key for ECDH should be a clear key generated by the default provider (i.e. by provider `bbb`). 

Not sure how to express that kind of logic in the property query string though....

Maybe something like `EC-KEYGEN#provider=bbb` ? Similar as specifying algorithms, one could specify the key gen operation of a certain key type as algorithm name in the property query, and thus select with which provider keys of a that type should be generated?

**Comment by t8m** [comment]
IMO the non-exportable private keys are handled already quite reasonably. Even currently for the keygen on a PKCS11 token you would just fetch the EC algorithm with propquery provider=pkcs11. You can then drop the propquery with later operations and it would still work even if the default propquery prefers the default provider i.e., ?provider=default

**Comment by levitte** [comment]
Oh...  I actually thought this was a nice idea for distinguishing EC keys on per curve basis.  From discussions here, that seems to be viewed differently somehow.  Not sure I understand why, though...

**Comment by ifranzki** [comment]
> You can then drop the propquery with later operations ...

The problem is that an SSL_CTX has exactly one property query that is set up at the very beginning. I have no chance to modify that later on. So if I use `provider=pkcs11` for the SSL_CTX it will attempt do do anything with the PKCS#11 provider, but will fail, because ECDH can not be done with PKCS#11 as long as we don't have symmetric opaque keys - aka SKEY - support.

So basically we need a way to tell SSL_CTX to use different providers for different things. This could be done through a smart way of specifying providers per algorithm in the one and only property query as discussed here.

**Comment by t8m** [comment]
Why would you ever want to use provider=pkcs11 with the SSL_CTX? You load the private key for the certificate separately and that can have a separate propquery (i.e., provider=pkcs11). If that key is unexportable it will be still usable for private key operations if you use ?provider=default with the SSL_CTX. So unless you need for example a different provider for the AEAD and a different provider for the key exchange, there should be no problem with the current APIs.

**Comment by ifranzki** [comment]
> If that key is unexportable it will be still usable for private key operations if you use ?provider=default with the SSL_CTX.

Why is this? Is it because the key originates from the from the PKCS#11 provider, so it prefers to use the algorithms from that provider as well? 

This is somehow contrary to the property queries.... Is that behavior anywhere documented? 

Even if the code behaves that way today, is this something I can trust that it stays that way forever? Instead of trusting an undocumented behavior I would rather like to use a documented way of selecting the providers, i.e. the property query. 

I would assume that a property query always wins over a binding to the provider where the key originates from. So even if a key originates from one provider, if the property query selects another provider for a certain algorithm, then the one provider selected by the property query must be used, and not the provider where the key originates from. Otherwise property queries would be totally unreliable. 

Assuming I use `?provider=default` for the SSL_CTX, and I have a key originated from the PKCS provider, and the default provider does support ECDSA for example, then OpenSSL MUST use the default provider for an ECDSA operation with this key. This is what the property query rules tell. 

If there is no property query (or an empty one), then I could accept that for a key that originates from a certain provider, the algorithm of that provider is preferred, but once a property query is present, then it must use that provider that matches the property query, and not take any shortcuts.

**Comment by t8m** [comment]
Not sure if this is documented but if it isn't it can be and should be. The unexportability of the key forces the operation to use the provider where the key is stored, if the property query allows that. Contrary to what you wrote above `?provider=default` does NOT tell that the default provider MUST be used. It tells that the default provider should be preferred for the operation, if possible. With unexportable keys, it is NOT possible to use the default provider, so the provider where the key is stored is used. Of course with `provider=default` query (without `?`) the operation will fail because such property query makes the default provider being required.

**Comment by t8m** [comment]
And yeah, this is expected to stay this way.

**Comment by ifranzki** [comment]
And how does the fetch logic know hat the key is non-extractable? It would only know when actually performing the export operation of the provider's KEYMGMT method to know it export works or not.
Can you point me to the places in the code that handles this ?

In any case, this behavior must be documented properly!

**Comment by paulidale** [comment]
> We should really summarize different use-cases separately because they might have different solutions. My gut feeling is that overly complicating the property query syntax would be a disaster from usability perspective.

Agreed.  I don't see this approach precluding curve selection for EC.  There are other possibilities which also need to be considered.

Note: I'm not suggesting any change to the property query syntax here, the change is an augmentation to the algorithm name.  I agree keeping the query syntax simple is worthwhile.

**Comment by t8m** [comment]
> And how does the fetch logic know hat the key is non-extractable? It would only know when actually performing the export operation of the provider's KEYMGMT method to know it export works or not. Can you point me to the places in the code that handles this ?
> 
> In any case, this behavior must be documented properly!

Yes, it does try to export the key and if that fails it implies it is not exportable.

**Comment by t8m** [comment]
**OTC: We should list the problems around fetching implementations first before we try to design solutions.**
